packages/core/snapshot/manager.go

```diff
@@ -0,0 +1,175 @@
+package snapshot
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/iotaledger/goshimmer/packages/core/epoch"
+	"github.com/iotaledger/goshimmer/packages/core/ledger"
+	"github.com/iotaledger/goshimmer/packages/core/notarization"
+	"github.com/iotaledger/goshimmer/packages/core/tangleold"
+	"github.com/iotaledger/goshimmer/packages/node/database"
+	"github.com/iotaledger/hive.go/core/generics/event"
+	"github.com/iotaledger/hive.go/core/kvstore"
+)
+
+const (
+	prefixSolidEntryPoint byte = iota
+)
+
+// Manager is the snapshot manager.
+type Manager struct {
+	tangle          *tangleold.Tangle
+	notarizationMgr *notarization.Manager
+
+	baseStore kvstore.KVStore
+	sepsLock  sync.RWMutex
+}
+
+// NewManager creates and returns a new snapshot manager.
+func NewManager(store kvstore.KVStore, t *tangleold.Tangle, nmgr *notarization.Manager) (new *Manager) {
+	new = &Manager{
+		tangle:          t,
+		notarizationMgr: nmgr,
+	}
+
+	new.baseStore = store
+
+	new.tangle.ConfirmationOracle.Events().BlockAccepted.Attach(event.NewClosure(func(e *tangleold.BlockAcceptedEvent) {
+		e.Block.ForEachParent(func(parent tangleold.Parent) {
+			index := parent.ID.EpochIndex
+			if index < e.Block.EI() {
+				new.insertSolidEntryPoint(parent.ID)
+			}
+		})
+	}))
+
+	new.tangle.ConfirmationOracle.Events().BlockOrphaned.Attach(event.NewClosure(func(event *tangleold.BlockAcceptedEvent) {
+		new.removeSolidEntryPoint(event.Block, event.Block.LatestConfirmedEpoch())
+	}))
+
+	return
+}
+
+// CreateSnapshot creates a snapshot file from node with a given name.
+func (m *Manager) CreateSnapshot(snapshotFileName string) (header *ledger.SnapshotHeader, err error) {
+	ecRecord, lastConfirmedEpoch, err := m.tangle.Options.CommitmentFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	// lock the entire ledger in notarization manager until the snapshot is created.
+	m.notarizationMgr.WriteLockLedger()
```

Comment:
We should lock before CommittmentFunc()

Reply:
There's a read lock in last confirmed and committable EI functions, so we can't put write lock before those. :thinking:

Reply:
I found a solution. :stuck_out_tongue: